File: Project/MyProject/MyProject/Source/MyProject/Private/BasicFrame/Libs/DataStruct/NoOrPriorityList/NoPriorityList/NoPriorityList.cpp
#include "MyProject.h"
#include "NoPriorityList.h"
#include "MacroDef.h"
#include "Ctx.h"
#include "LogSys.h"
#include "LogTypeId.h"

MY_BEGIN_NAMESPACE(MyNS)

NoPriorityList::NoPriorityList()
{
	//this->mNoPriorityProcessObjectList = new MList<INoOrPriorityObject>();
	this->mIsSpeedUpFind = false;
}

void NoPriorityList::init()
{

}

void NoPriorityList::dispose()
{
	//if(nullptr != this->mNoPriorityProcessObjectList)
	//{
		this->mNoPriorityProcessObjectList.dispose();
		//this->mNoPriorityProcessObjectList = nullptr;
	//}
	//if (nullptr != this->mDic)
	//{
		this->mDic.dispose();
		//this->mDic = nullptr;
	//}

	this->clear();
}

void NoPriorityList::setIsSpeedUpFind(bool value)
{
	this->mIsSpeedUpFind = value;

	//if (this->mIsSpeedUpFind)
	//{
	//	this->mDic = new MDictionary<INoOrPriorityObject, int>();
	//}
}

void NoPriorityList::setIsOpKeepSort(bool value)
{

}

void NoPriorityList::clear()
{
	this->mNoPriorityProcessObjectList.clear();

	if(this->mIsSpeedUpFind)
	{
		this->mDic.clear();
	}
}

int NoPriorityList::count()
{
	return this->mNoPriorityProcessObjectList.count();
}

INoOrPriorityObject* NoPriorityList::get(int index)
{
	INoOrPriorityObject* ret = nullptr;

	if(index < this->count())
	{
		ret = this->mNoPriorityProcessObjectList.get(index);
	}

	return ret;
}

bool NoPriorityList::contains(INoOrPriorityObject* item)
{
	bool ret = false;

	if (nullptr != item)
	{
		if (this->mIsSpeedUpFind)
		{
			ret = this->mDic.containsKey(item);
		}
		else
		{
			int index = 0;
			int listLen = this->mNoPriorityProcessObjectList.count();

			while (index < listLen)
			{
				if (item == this->mNoPriorityProcessObjectList.get(index))
				{
					ret = true;
					break;
				}

				++index;
			}
		}
	}
	else
	{
		if (MacroDef::ENABLE_LOG)
		{
			GLogSys->log("NoPriorityList::Contains, failed", LogTypeId::eLogNoPriorityListCheck);
		}
	}

	return ret;
}

void NoPriorityList::removeAt(int index)
{
	if (this->mIsSpeedUpFind)
	{
		this->_effectiveRemove(this->mNoPriorityProcessObjectList[index]);
	}
	else
	{
		this->mNoPriorityProcessObjectList.removeAt(index);
	}
}

int NoPriorityList::getIndexByNoPriorityObject(INoOrPriorityObject* priorityObject)
{
	int retIndex = -1;

	int index = 0;
	int listLen = this->mNoPriorityProcessObjectList.count();

	while (index < listLen)
	{
		if (this->mNoPriorityProcessObjectList.get(index) == priorityObject)
		{
			retIndex = index;
			break;
		}

		++index;
	}

	return retIndex;
}

int NoPriorityList::getIndexByNoOrPriorityObject(INoOrPriorityObject* priorityObject)
{
	return this->getIndexByNoPriorityObject(priorityObject);
}

void NoPriorityList::addNoPriorityObject(INoOrPriorityObject* noPriorityObject)
{
	if (nullptr != noPriorityObject)
	{
		if (!this->contains(noPriorityObject))
		{
			this->mNoPriorityProcessObjectList.add(noPriorityObject);

			if (this->mIsSpeedUpFind)
			{
				this->mDic.add(noPriorityObject, this->mNoPriorityProcessObjectList.count() - 1);
			}
		}
	}
	else
	{
		if (MacroDef::ENABLE_LOG)
		{
			GLogSys->log("NoPriorityList::addNoPriorityObject, failed", LogTypeId::eLogNoPriorityListCheck);
		}
	}
}

void NoPriorityList::removeNoPriorityObject(INoOrPriorityObject* noPriorityObject)
{
	if (nullptr != noPriorityObject)
	{
		if (this->contains(noPriorityObject))
		{
			if (this->mIsSpeedUpFind)
			{
				this->_effectiveRemove(noPriorityObject);
			}
			else
			{
				int index = this->getIndexByNoPriorityObject(noPriorityObject);

				if (-1 != index)
				{
					this->mNoPriorityProcessObjectList.removeAt(index);
				}
			}
		}
	}
	else
	{
		if (MacroDef::ENABLE_LOG)
		{
			GLogSys->log("NoPriorityList::addNoPriorityObject, failed", LogTypeId::eLogNoPriorityListCheck);
		}
	}
}

void NoPriorityList::addNoOrPriorityObject(INoOrPriorityObject* noPriorityObject, float priority)
{
	this->addNoPriorityObject(noPriorityObject);
}

void NoPriorityList::removeNoOrPriorityObject(INoOrPriorityObject* noPriorityObject)
{
	this->removeNoPriorityObject(noPriorityObject);
}

// 快速移除元素
bool NoPriorityList::_effectiveRemove(INoOrPriorityObject* item)
{
	bool ret = false;

	if (this->mDic.containsKey(item))
	{
		ret = true;

		int index = this->mDic[item];
		this->mDic.remove(item);

		if (index == this->mNoPriorityProcessObjectList.count() - 1)    // 如果是最后一个元素，直接移除
		{
			this->mNoPriorityProcessObjectList.removeAt(index);
		}
		else
		{
			this->mNoPriorityProcessObjectList.set(index, this->mNoPriorityProcessObjectList.get(this->mNoPriorityProcessObjectList.count() - 1));
			this->mNoPriorityProcessObjectList.removeAt(this->mNoPriorityProcessObjectList.count() - 1);
			this->mDic.add(this->mNoPriorityProcessObjectList.get(index), index);
		}
	}

	return ret;
}

void NoPriorityList::_updateIndex(int index)
{
	int listLen = this->mNoPriorityProcessObjectList.count();

	while (index < listLen)
	{
		this->mDic.add(this->mNoPriorityProcessObjectList.get(index), index);

		++index;
	}
}

MY_END_NAMESPACE

File: Project/MyProject/MyProject/Content/MyAsset/Lua/MyLua/Libs/DataStruct/NoOrPriorityList/NoPriorityList/NoPriorityList.lua
MLoader("MyLua.Libs.Core.GlobalNS");
MLoader("MyLua.Libs.Core.Class");
MLoader("MyLua.Libs.Core.GObject");

MLoader("MyLua.Libs.DataStruct.MList");
MLoader("MyLua.Libs.DataStruct.MDictionary");

-- 优先级队列
local M = GlobalNS.Class(GlobalNS.GObject);
M.clsName = "NoPriorityList";
GlobalNS[M.clsName] = M;

function M:ctor()
	self.mNoPriorityProcessObjectList = GlobalNS.new(GlobalNS.MList);
	self.mIsSpeedUpFind = false;
end

function M:dtor()

end

function M:setIsSpeedUpFind(value)
	self.mIsSpeedUpFind = value;

	if (self.mIsSpeedUpFind) then
		self.mDic = GlobalNS.new(GlobalNS.MDictionary);
	end
end

function M:setIsOpKeepSort(value)

end

function M:clear()
	self.mNoPriorityProcessObjectList:clear();

	if(self.mIsSpeedUpFind) then
		self.mDic:clear();
	end
end

function M:count()
	return self.mNoPriorityProcessObjectList:count();
end

function M:get(index)
	local ret = nil;

	if(index < self:count()) then
		ret = self.mNoPriorityProcessObjectList:get(index);
	end

	return ret;
end

function M:contains(item)
	local ret = false;

	if (nil ~= item) then
		if (self.mIsSpeedUpFind) then
			ret = self.mDic:containsKey(item);
		else
			local index = 0;
			local listLen = self.mNoPriorityProcessObjectList:count();

			while (index < listLen) do
				if (item == self.mNoPriorityProcessObjectList:get(index)) then
					ret = true;
					break;
				end

				index = index + 1;
			end
		end
	else
		if (MacroDef.ENABLE_LOG) then
			Ctx.mLogSys:log("NoPriorityList::Contains, failed", GlobalNS.LogTypeId.eLogNoPriorityListCheck);
		end
	end

	return ret;
end

function M:removeAt(index)
	if (self.mIsSpeedUpFind) then
		self:effectiveRemove(self.mNoPriorityProcessObjectList:at(index));
	else
		self.mNoPriorityProcessObjectList:removeAt(index);
	end
end

function M:getIndexByNoPriorityObject(priorityObject)
	local retIndex = -1;

	local index = 0;
	local listLen = self.mNoPriorityProcessObjectList:count();

	while (index < listLen) do
		if (self.mNoPriorityProcessObjectList:get(index) == priorityObject) then
			retIndex = index;
			break;
		end

		index = index + 1;
	end

	return retIndex;
end

function M:getIndexByNoOrPriorityObject(priorityObject)
	return self:getIndexByNoPriorityObject(priorityObject);
end

function M:addNoPriorityObject(noPriorityObject)
	if (nil ~= noPriorityObject) then
		if (not self:contains(noPriorityObject)) then
			self.mNoPriorityProcessObjectList:add(noPriorityObject);

			if (self.mIsSpeedUpFind) then
				self.mDic:add(noPriorityObject, self.mNoPriorityProcessObjectList:count() - 1);
			end
		end
	else
		if (MacroDef.ENABLE_LOG) then
			Ctx.mLogSys:log("NoPriorityList::addNoPriorityObject, failed", GlobalNS.LogTypeId.eLogNoPriorityListCheck);
		end
	end
end

function M:removeNoPriorityObject(noPriorityObject)
	if (nil ~= noPriorityObject) then
		if (self:contains(noPriorityObject)) then
			if (self.mIsSpeedUpFind) then
				self:effectiveRemove(noPriorityObject);
			else
				local index = self:getIndexByNoPriorityObject(noPriorityObject);

				if (-1 ~= index) then
					self.mNoPriorityProcessObjectList:removeAt(index);
				end
			end
		end
	else
		if (MacroDef.ENABLE_LOG) then
			Ctx.mLogSys:log("NoPriorityList::addNoPriorityObject, failed", GlobalNS.LogTypeId.eLogNoPriorityListCheck);
		end
	end
end

function M:addNoOrPriorityObject(noPriorityObject, priority)
	self:addNoPriorityObject(noPriorityObject);
end

function M:removeNoOrPriorityObject(noPriorityObject)
	self:removeNoPriorityObject(noPriorityObject);
end

-- 快速移除元素
function M:effectiveRemove(item)
	local ret = false;

	if (self.mDic:containsKey(item)) then
		ret = true;

		local idx = self.mDic:value(item);
		self.mDic:remove(item);

		if (idx == self.mNoPriorityProcessObjectList:count() - 1) then   -- 如果是最后一个元素，直接移除
			self.mNoPriorityProcessObjectList:removeAt(idx);
		else
			self.mNoPriorityProcessObjectList:set(idx, self.mNoPriorityProcessObjectList:get(self.mNoPriorityProcessObjectList:count() - 1));
			self.mNoPriorityProcessObjectList:removeAt(self.mNoPriorityProcessObjectList:count() - 1);
			self.mDic:add(self.mNoPriorityProcessObjectList:get(idx), idx);
		end
	end

	return ret;
end

function M:updateIndex(idx)
	local listLen = self.mNoPriorityProcessObjectList:count();

	while (idx < listLen) do
		self.mDic:add(self.mNoPriorityProcessObjectList:get(idx), idx);

		idx = idx + 1;
	end
end

return M;